internal/pass/sync/sync.go

```diff
@@ -0,0 +1,178 @@
+package sync
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/twofas/2fas-server/internal/common/logging"
+	"github.com/twofas/2fas-server/internal/pass/connection"
+	"github.com/twofas/2fas-server/internal/pass/sign"
+)
+
+type Syncing struct {
+	store   store
+	signSvc *sign.Service
+}
+
+type store interface {
+	RequestSync(fmtToken string)
+	ConfirmSync(fmtToken string) bool
+	IsSyncCofirmed(fmtToken string) bool
+}
+
+func NewPairingApp(signService *sign.Service, fakeMobilePush bool) *Syncing {
+	return &Syncing{
+		store:   NewMemoryStore(),
+		signSvc: signService,
+	}
+}
+
+const (
+	pairingTokenValidityDuration = 3 * time.Minute
+)
+
+type ConfigureBrowserExtensionRequest struct {
+	ExtensionID string `json:"extension_id"`
+}
+
+type ConfigureBrowserExtensionResponse struct {
+	BrowserExtensionPairingToken string `json:"browser_extension_pairing_token"`
+	ConnectionToken              string `json:"connection_token"`
+}
+
+type ExtensionWaitForConnectionInput struct {
+	ResponseWriter http.ResponseWriter
+	HttpReq        *http.Request
+}
+
+type RequestSyncResponse struct {
+	BrowserExtensionProxyToken string `json:"browser_extension_proxy_token"`
+	Status                     string `json:"status"`
+}
+
+type MobileSyncPayload struct {
+	MobileSyncToken string `json:"mobile_sync_token"`
+}
+
+func (p *Syncing) ServeSyncingRequestWS(w http.ResponseWriter, r *http.Request, fcmToken string) error {
+	log := logging.WithField("fcm_token", fcmToken)
+	conn, err := connection.Upgrade(w, r)
+	if err != nil {
+		return fmt.Errorf("failed to upgrade connection: %w", err)
+	}
+	defer conn.Close()
+
+	log.Infof("Starting sync request WS for %q", fcmToken)
+	p.requestSync(r.Context(), fcmToken)
+
+	if syncDone := p.isSyncConfirmed(r.Context(), fcmToken); syncDone {
+		if err := p.sendTokenAndCloseConn(fcmToken, conn); err != nil {
+			log.Errorf("Failed to send token: %v", err)
```

Comment:
why don't you return error instread of logging it here?

Reply:
Because it's useless. Caller can't do anything with the error (i.e. return error code from the handler). We can return error from this method and then try to distinguish in the caller if this is pre or post upgrade. But maybe this is simpler?